core/psh/ping.c

```diff
@@ -0,0 +1,279 @@
+		fprintf(stderr, "ping: Can't open socket!\n");
+		return -EIO;
+	}
+
+	if (setsockopt(fd, IPPROTO_IP, IP_TTL, &ping_common.ttl, sizeof(ping_common.ttl)) != 0) {
+		close(fd);
+		fprintf(stderr, "ping: Can't set TTL!\n");
+		return -EIO;
+	}
+
+	tv.tv_sec = ping_common.timeout / 1000;
+	tv.tv_usec = (ping_common.timeout % 1000) * 1000;
+	if (setsockopt(fd, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv)) != 0) {
+		close(fd);
+		fprintf(stderr, "ping: Can't set socket timeout!\n");
+		return -EIO;
+	}
+
+	return fd;
+}
+
+
+static void psh_ping_reqInit(uint8_t *data, int len)
+{
+	struct icmphdr *hdr = (struct icmphdr *) data;
+	int i;
+
+	hdr->type = ICMP_ECHO;
+	hdr->un.echo.id = htons(getpid());
+
+	for (i = 0; i < len - sizeof(struct icmphdr); i++)
+		data[sizeof(struct icmphdr) + i] = (uint8_t) i;
+}
+
+
+static int psh_ping_send(int fd, uint8_t *data, int data_len)
+{
+	struct icmphdr *hdr = (struct icmphdr *) data;
+
+	hdr->un.echo.sequence = htons(ping_common.msg_seq++);
+	hdr->checksum = 0;
+	hdr->checksum = htons(psh_ping_chksum(data, data_len));
+
+	if (sendto(fd, data, data_len, 0, (struct sockaddr *) &ping_common.r_addr,
+	           sizeof(ping_common.r_addr)) != data_len)
+		return -1;
+
+	return 0;
+}
+
+
+static int psh_ping_recv(int fd, uint8_t *data, int data_len, char *infostr, int infostrlen)
+{
+	struct sockaddr_in r_sin;
+	socklen_t r_len;
+	struct iphdr *iphdr;
+	struct icmphdr *icmphdr;
+	int bytes;
+	char addrstr[INET_ADDRSTRLEN];
+	uint16_t chksum;
+
+	if ((bytes = recvfrom(fd, data, data_len, 0, (struct sockaddr *) &r_sin, &r_len)) <= 0) {
+		if (errno == EAGAIN || errno == EWOULDBLOCK)
+			fprintf(stderr, "Host unreachable\n");
+		else
+			fprintf(stderr, "ping: Fail to receive packet on socket!\n");
+		return -1;
+	}
+
+	if (bytes < sizeof(struct iphdr) + sizeof(struct icmphdr)) {
+		fprintf(stderr, "ping: Received msg too short (%d)!\n", bytes);
+		return -1;
+	}
+
+	if (inet_ntop(ping_common.af, &r_sin.sin_addr, addrstr, sizeof(addrstr)) == NULL) {
+		fprintf(stderr, "ping: Invalid address received!\n");
+		return -1;
+	}
+
+	if (memcmp(&r_sin.sin_addr, &ping_common.r_addr.sin_addr, sizeof(r_sin.sin_addr))) {
+		fprintf(stderr, "ping: Response from invalid address: %s!\n", addrstr);
+		return -1;
+	}
+
+	iphdr = (struct iphdr *) data;
+	icmphdr = (struct icmphdr *) (data + sizeof(struct iphdr));
+	chksum = ntohs(icmphdr->checksum);
+	icmphdr->checksum = 0;
+
+	if (psh_ping_chksum(data, bytes) != chksum) {
+		fprintf(stderr, "ping: Response invalid checksum!\n", addrstr);
+		return -1;
+	}
+
+	return snprintf(infostr, infostrlen, "%d bytes received from %s: ttl=%u icmp_seq=%u",
+	                bytes - sizeof(struct iphdr), addrstr, iphdr->ttl, ntohs(icmphdr->un.echo.sequence));
+}
+
+
+int psh_ping(int argc, char **argv)
+{
+	uint8_t msg_req[PING_PKT_SIZE] = { 0 };
+	uint8_t msg_resp[PING_PKT_SIZE * 2] = { 0 };
+	char infostr[80];
+	unsigned long sent_time, elapsed;
```

Comment:
You're right, it would be more appropriate to use `time_t`.

Reply:
@MaciejPurski, according to our coding rules please don't use underline in variable names. The convention of defining macros is to start them with a noun naming the purpose/desired usage of particular definition. For example `msg_req` array should be defined in the code as `req[SIZE_PING]` or even `msgReq[SIZE_PKTPING]`.

Reply:
@MaciejPurski, I don't understand the reason of defining size of ping packet using macro. You could do this by using numbers and using `sizeof()` operator when you need to obtain the size of this variable. In your code there is only one line (254) in which you refer to this macro. It could be very easily changed with `sizeof()`.

Reply:
Done.